File: app/src/main/java/com/kickstarter/services/apiresponses/ErrorEnvelope.java
package com.kickstarter.services.apiresponses;

import android.os.Parcelable;

import com.kickstarter.libs.qualifiers.AutoGson;
import com.kickstarter.libs.utils.ListUtils;
import com.kickstarter.services.ApiException;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringDef;
import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class ErrorEnvelope implements Parcelable {
  public abstract @Nullable List<String> errorMessages();
  public abstract int httpCode();
  public abstract @Nullable String ksrCode();
  public abstract @Nullable FacebookUser facebookUser();

  @AutoGson
  @AutoParcel
  public static abstract class FacebookUser implements Parcelable {
    public abstract long id();
    public abstract String name();
    public abstract String email();

    @AutoParcel.Builder
    public static abstract class Builder {
      public abstract Builder id(long __);
      public abstract Builder name(String __);
      public abstract Builder email(String __);
      public abstract FacebookUser build();
    }

    public static Builder builder() {
      return new AutoParcel_ErrorEnvelope_FacebookUser.Builder();
    }

    public abstract Builder toBuilder();
  }

  @AutoParcel.Builder
  public abstract static class Builder {
    public abstract Builder errorMessages(List<String> __);
    public abstract Builder httpCode(int __);
    public abstract Builder ksrCode(String __);
    public abstract Builder facebookUser(FacebookUser __);
    public abstract ErrorEnvelope build();
  }

  public static Builder builder() {
    return new AutoParcel_ErrorEnvelope.Builder();
  }

  public abstract Builder toBuilder();

  public static final String CONFIRM_FACEBOOK_SIGNUP = "confirm_facebook_signup";
  public static final String INVALID_XAUTH_LOGIN = "invalid_xauth_login";
  public static final String TFA_FAILED = "tfa_failed";
  public static final String TFA_REQUIRED = "tfa_required";
  public static final String MISSING_FACEBOOK_EMAIL = "missing_facebook_email";
  public static final String FACEBOOK_INVALID_ACCESS_TOKEN = "facebook_invalid_access_token";
  public static final String UNAUTHORIZED = "unauthorized";

  @StringDef({INVALID_XAUTH_LOGIN, TFA_FAILED, TFA_REQUIRED})
  @Retention(RetentionPolicy.SOURCE)
  public @interface ErrorCode {}

  /**
   * Tries to extract an {@link ErrorEnvelope} from an exception, and if it
   * can't returns null.
   */
  public static @Nullable ErrorEnvelope fromThrowable(final @NonNull Throwable t) {
    if (t instanceof ApiException) {
      final ApiException exception = (ApiException) t;
      return exception.errorEnvelope();
    }

    return null;
  }

  public boolean isConfirmFacebookSignupError() {
    return CONFIRM_FACEBOOK_SIGNUP.equals(ksrCode());
  }

  public boolean isInvalidLoginError() {
    return INVALID_XAUTH_LOGIN.equals(ksrCode());
  }

  public boolean isTfaRequiredError() {
    return TFA_REQUIRED.equals(ksrCode());
  }

  public boolean isTfaFailedError() {
    return TFA_FAILED.equals(ksrCode());
  }

  public boolean isMissingFacebookEmailError() {
    return MISSING_FACEBOOK_EMAIL.equals(ksrCode());
  }

  public boolean isFacebookInvalidAccessTokenError() {
    return FACEBOOK_INVALID_ACCESS_TOKEN.equals(ksrCode());
  }

  public boolean isUnauthorizedError() {
    return UNAUTHORIZED.equals(ksrCode());
  }

  /*
    When logging in the only two possible errors are INVALID_XAUTH_LOGIN
    and TFA_REQUIRED, so we consider anything else an unknown error.
   */
  public boolean isGenericLoginError() {
    return
      !INVALID_XAUTH_LOGIN.equals(ksrCode()) &&
        !TFA_REQUIRED.equals(ksrCode());
  }

  /**
   * Returns the first error message available, or `null` if there are none.
   */
  public @Nullable String errorMessage() {
    if (errorMessages() == null) {
      return null;
    } else {
      return ListUtils.first(errorMessages());
    }
  }
}

File: app/src/main/java/com/kickstarter/services/apiresponses/ErrorEnvelope.kt
package com.kickstarter.services.apiresponses

import android.os.Parcelable
import androidx.annotation.StringDef
import com.kickstarter.services.ApiException
import kotlinx.parcelize.Parcelize
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy

@Parcelize
class ErrorEnvelope private constructor(
    private val errorMessages: List<String> = emptyList(),
    private val httpCode: Int = 0,
    private val ksrCode: String = "",
    private val facebookUser: FacebookUser? = null // - Re-evaluate on https://kickstarter.atlassian.net/browse/MBL-815 the default value for this specific field
) : Parcelable {
    fun errorMessages() = this.errorMessages
    fun httpCode() = this.httpCode
    fun ksrCode() = this.ksrCode
    fun facebookUser() = this.facebookUser

    @Parcelize
    class FacebookUser private constructor(
        private val id: Long = 0L,
        private val name: String = "",
        private val email: String = ""
    ) : Parcelable {
        fun id() = this.id
        fun name() = this.name
        fun email() = this.email

        @Parcelize
        data class Builder(
            private var id: Long = 0L,
            private var name: String = "",
            private var email: String = ""
        ) : Parcelable {
            fun id(id: Long?) = apply { this.id = id ?: 0L }
            fun name(name: String?) = apply { this.name = name ?: "" }
            fun email(email: String?) = apply { this.email = email ?: "" }
            fun build() = FacebookUser(
                id = id,
                name = name,
                email = email
            )
        }

        fun toBuilder() = Builder(
            id = id,
            name = name,
            email = email
        )

        override fun equals(other: Any?): Boolean {
            var equals = super.equals(other)
            if (other is FacebookUser) {
                equals = id() == other.id() &&
                    name() == other.name() &&
                    email() == other.email()
            }
            return equals
        }

        companion object {
            @JvmStatic
            fun builder(): Builder {
                return Builder()
            }
        }
    }

    data class Builder(
        private var errorMessages: List<String> = emptyList(),
        private var httpCode: Int = 0,
        private var ksrCode: String = "",
        private var facebookUser: FacebookUser? = null
    ) {
        fun errorMessages(errorMessages: List<String?>?) = apply { errorMessages?.let { this.errorMessages = it.filterNotNull() } ?: emptyList<String>() }
        fun httpCode(httpCode: Int?) = apply { this.httpCode = httpCode ?: 0 }
        fun ksrCode(ksrCode: String?) = apply { this.ksrCode = ksrCode ?: "" }
        fun facebookUser(facebookUser: FacebookUser?) = apply { this.facebookUser = facebookUser }
        fun build() = ErrorEnvelope(
            errorMessages = errorMessages,
            httpCode = httpCode,
            ksrCode = ksrCode,
            facebookUser = facebookUser
        )
    }

    fun toBuilder() = Builder(
        errorMessages = errorMessages,
        httpCode = httpCode,
        ksrCode = ksrCode,
        facebookUser = facebookUser
    )

    @StringDef(INVALID_XAUTH_LOGIN, TFA_FAILED, TFA_REQUIRED)
    @Retention(RetentionPolicy.SOURCE)
    annotation class ErrorCode

    val isConfirmFacebookSignupError: Boolean
        get() = CONFIRM_FACEBOOK_SIGNUP == ksrCode()
    val isInvalidLoginError: Boolean
        get() = INVALID_XAUTH_LOGIN == ksrCode()
    val isTfaRequiredError: Boolean
        get() = TFA_REQUIRED == ksrCode()
    val isTfaFailedError: Boolean
        get() = TFA_FAILED == ksrCode()
    val isMissingFacebookEmailError: Boolean
        get() = MISSING_FACEBOOK_EMAIL == ksrCode()
    val isFacebookInvalidAccessTokenError: Boolean
        get() = FACEBOOK_INVALID_ACCESS_TOKEN == ksrCode()
    val isUnauthorizedError: Boolean
        get() = UNAUTHORIZED == ksrCode()

    /*
    When logging in the only two possible errors are INVALID_XAUTH_LOGIN
    and TFA_REQUIRED, so we consider anything else an unknown error.
   */
    val isGenericLoginError: Boolean
        get() = INVALID_XAUTH_LOGIN != ksrCode() &&
            TFA_REQUIRED != ksrCode()

    /**
     * Returns the first error message available, or "" if there are none.
     */
    fun errorMessage(): String = errorMessages()?.firstOrNull() ?: ""

    override fun equals(other: Any?): Boolean {
        var equals = super.equals(other)
        if (other is ErrorEnvelope) {
            equals = errorMessages() == other.errorMessages() &&
                httpCode() == other.httpCode() &&
                ksrCode() == other.ksrCode() &&
                facebookUser() == other.facebookUser()
        }
        return equals
    }

    companion object {
        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }

        const val CONFIRM_FACEBOOK_SIGNUP = "confirm_facebook_signup"
        const val INVALID_XAUTH_LOGIN = "invalid_xauth_login"
        const val TFA_FAILED = "tfa_failed"
        const val TFA_REQUIRED = "tfa_required"
        const val MISSING_FACEBOOK_EMAIL = "missing_facebook_email"
        const val FACEBOOK_INVALID_ACCESS_TOKEN = "facebook_invalid_access_token"
        const val UNAUTHORIZED = "unauthorized"

        /**
         * Tries to extract an [ErrorEnvelope] from an exception, and if it
         * can't returns Empty [ErrorEnvelope].
         */
        @JvmStatic
        fun fromThrowable(t: Throwable): ErrorEnvelope {
            if (t is ApiException) {
                return t.errorEnvelope()
            }

            return ErrorEnvelope.builder().errorMessages(listOf(t.message ?: "")).build()
        }
    }
}
